gemini/sample-apps/llamaindex-rag/src/app/main.py

```diff
@@ -0,0 +1,484 @@
+from src.app.shared_state import index_manager, prompts
+
+
+def get_index_manager():
+    logger.info(index_manager.base_index.docstore)
+    logger.info(index_manager.firestore_db_name)
+    logger.info(index_manager.firestore_namespace)
+    return index_manager
+
+
+def get_prompts():
+    return prompts
+
+
+creds = google.auth.default()[0]
+request = google.auth.transport.requests.Request()
+
+app = FastAPI()
+
+
+@app.get("/")
+def root():
+    return {"hello world": "hello world"}
+
+
+@app.get("/get_all_prompts")
+async def get_all_prompts(prompts=Depends(get_prompts)):
+    return prompts.to_dict()
+
+
+@app.post("/update_prompt")
+async def update_prompt(prompt_update: PromptUpdate, prompts=Depends(get_prompts)):
+    try:
+        prompts.update(prompt_update.prompt_name, prompt_update.new_content)
+        return {"message": f"Prompt {prompt_update.prompt_name} updated successfully"}
+    except ValueError as e:
+        raise HTTPException(status_code=400, detail=str(e))
+
+
+@app.post("/list_vector_search_indexes")
+async def list_vector_search_indexes(
+    qa_or_base: dict, index_manager=Depends(get_index_manager)
+):
+    index_list = aiplatform.MatchingEngineIndex.list()
+    index_list = [i.display_name for i in index_list]
+    if qa_or_base["qa_or_base"] == "qa":
+        active_index_name = index_manager.qa_index_name
+    elif qa_or_base["qa_or_base"] == "base":
+        active_index_name = index_manager.base_index_name
+    if active_index_name is None:
+        index_list.insert(0, None)
+    else:
+        index_list.remove(active_index_name)
+        index_list.insert(0, active_index_name)
+    return index_list
+
+
+@app.post("/list_vector_search_endpoints")
+async def list_vector_search_endpoints(
+    qa_or_base: dict, index_manager=Depends(get_index_manager)
+):
+    endpoint_list = aiplatform.MatchingEngineIndexEndpoint.list()
+    endpoint_list = [e.display_name for e in endpoint_list]
+    if qa_or_base["qa_or_base"] == "qa":
+        active_endpoint_name = index_manager.qa_endpoint_name
+    elif qa_or_base["qa_or_base"] == "base":
+        active_endpoint_name = index_manager.base_endpoint_name
+    if active_endpoint_name is None:
+        endpoint_list.insert(0, None)
+    else:
+        endpoint_list.remove(active_endpoint_name)
+        endpoint_list.insert(0, active_endpoint_name)
+    return endpoint_list
+
+
+@app.get("/list_firestore_databases")
+async def list_firestore_databases(index_manager=Depends(get_index_manager)):
+    active_db_name = index_manager.firestore_db_name
+    client_options = ClientOptions(api_endpoint="firestore.googleapis.com")
+    client = firestore_admin_v1.FirestoreAdminClient(client_options=client_options)
+    parent = f"projects/{index_manager.project_id}"
+    databases = []
+
+    try:
+        # List all databases in the project
+        for database in client.list_databases(parent=parent).databases:
+            databases.append(database.name.split("/")[-1])
```

Comment:
There's a method on the client to parse those full database names and get the end part by name.

Reply:
I wasn't able to find this method